neuralpp/symbolic/general_normalizer.py

```diff
@@ -0,0 +1,121 @@
+            normalized_body = _normalize(body, context & constraint)
+            match normalized_body:
+                case FunctionApplication(function=Constant(value=functions.conditional),
+                                         arguments=[condition, then, else_]):
+                    if condition.contains(index):
+                        return _normalize(
+                            operation(BasicQuantifierExpression(operation, index, constraint & condition, then),
+                                      BasicQuantifierExpression(operation, index, constraint & ~condition, else_)),
+                            context)
+                    else:
+                        return if_then_else(condition,
+                                            _normalize(BasicQuantifierExpression(operation, index, constraint, then),
+                                                       context & condition),
+                                            _normalize(BasicQuantifierExpression(operation, index, constraint, else_),
+                                                       context & ~condition))
+                case _:
+                    return _eliminate(operation, index, constraint, normalized_body, context)
+
+
+simplifier = ContextSimplifier()
+
+
+def _eliminate(operation: AbelianOperation, index: Variable, constraint: Context, body: Expression,
+               context: Z3SolverExpression) -> Expression:
+    """
+    Eliminates all quantifiers by doing the "summation" (in a broader sense).
+    Currently, a very naive placeholder.
+    In particular, we allow eliminate() to accept body and return expression that contains quantifier,
+    thus it is a "partial/best-effort elimination".
+    In particular, we expect body and result to be normalized quantifiers-at-leaves.
+
+    Future work: more complicated elimination algorithm, which actually tries to `eliminate` quantifiers.
+    In general 2 directions for improvement:
+    1. supports more operations (add, multiply, and, or, ...)
+    2. supports multiple intervals & complicated constraints (e.g, 1 <= x <= 100, x != y)
+    """
+    if context.is_known_to_imply(~constraint):
+        return operation.identity
+    return BasicQuantifierExpression(operation, index, constraint, simplifier.simplify(body, context & constraint))
+
+
+def _normalize_function_application(function: Expression,
+                                    arguments: List[Expression],
+                                    context: Z3SolverExpression, i: int = 0) -> Expression:
+    if i >= len(arguments):
+        return function(*arguments)
+    n_i = _normalize(arguments[i], context)
+    arguments[i] = n_i
+    return _move_down_and_normalize(function, arguments, context, i)
+
+
+def _move_down_and_normalize(function: Expression,
+                             arguments: List[Expression],
+                             context: Z3SolverExpression, i: int) -> Expression:
+    # assume i < len(arguments);
```

Comment:
I feel these can stay in comments since they're requirements on an internal function, so these assumptions are more intended to help readers of the code understand the function (than posing requirements on the argument)? Or I think even if we were to add `assert`, they should be for debug only.